### ISSUE TITLE
Ideas

### ISSUE THREAD

**Issue by ankane** [ankane/lockbox#1]
0.7.0

- [x] Drop support for Active Record < 5.2 and Ruby < 2.6
- [x] Raise an error for unsupported versions of Active Record and CarrierWave

1.0.0 (breaking)

- [ ] Add `binary` option to replace `encode` and encode by default for `Lockbox.new`
- [ ] [maybe] Decode to UTF-8 unless `binary: true` - `utf8` branch
- [ ] Don't encode in Base64 for `binary` database fields (if simple to implement)
- [ ] [maybe] Create new blobs when blobs are attached without `encrypted` flag

Considering

- [ ] `Lockbox::Redis` and `Lockbox::Dalli` for cache stores (probably better as separate gem) - `cache_stores` branch - or `Lockbox::Cache::Store` for Active Support cache store
- [ ] Prefer `encrypt_attribute` and `decrypt_attribute` over `generate_attribute_ciphertext` and `decrypt_attribute_ciphertext`
- [ ] Don't return virtual attributes in attribute methods (`attributes`, `attribute_names`, `has_attribute?`) unless the ciphertext attribute is present
- [ ] Use Fiddle for Libsodium - `libsodium` branch
- [ ] Warn (and eventually throw error) if the master key is passed to `Lockbox.new`
- [ ] Require `allow_empty` option to encrypt empty string without padding
- [ ] Encrypt empty strings in database fields - `model_empty_string` branch
- [ ] Add support for encrypted Active Storage service (can wrap any other service) - more useful in 6.1+ since multiple services are supported (blocked since encryption needs to happen before checksum is computed)
- [ ] Store the encryption version to make it easy to see which data has been rotated and avoid trying multiple keys. Could be done in an optional new field (`email_ciphertext_version`, `license_version`, blob `metadata`) or directly in the ciphertext (needs to work for files/binary data)
- [ ] Prefer `encrypts` over `encrypt` for CarrierWave
- [ ] Default padding for encoded strings to reduce data leakage (cons: less standard, slightly more space)
- [ ] Prefer `ActiveSupport.on_load(:action_text_rich_text) { ActionText::RichText.encrypts :body }` over `Lockbox.encrypts_action_text_body` (more code but less magic)
- [ ] Add `associated_data` option to models
- [ ] Add `pretty_print` method (similar to `inspect`)

On hold

- [ ] Support for streaming encryption (probably not needed) - `streaming` branch
- [ ] Better support for KMS (store key in data/metadata instead of DB) - `kms_encrypt` branch
- [ ] Shrine support - `shrine` branch - WIP

**Comment by cj** [comment]
@ankane I'd be really interested in streaming encryption.  Let me know if I can help.

**Comment by ankane** [comment]
Hey @cj, sorry for the long delay. If this is still interesting, can you explain a bit about your use case? (I'm a bit hesitant to add since most of the time files fit into memory).

**Comment by hovancik** [comment]
I would love to see per user encryption based on their password (password-derived key)

**Comment by ankane** [comment]
Hey @hovancik, can you explain more about your use case and how you'd like it to work?

**Comment by hovancik** [comment]
@ankane  I would like to use so called Password Based Encryption, where usually encryption key is generated from user's password. [This](https://en.wikipedia.org/wiki/PBKDF2) is usually used for that. 

Example users of this are ie Firefox accounts, Lastpass master password etc.

**Comment by ankane** [comment]
Password-based encryption should happen on the client (here's the [LastPass architecture](https://www.lastpass.com/enterprise/security)), so I'm not sure it's common to be implemented in Ruby. That being said, you can already use PBKDF2 or Argon2 (more secure) for key derivation and then pass the key to Lockbox.

**Comment by sgriff44** [comment]
@ankane, Loved how easy this was to setup in an existing application - thank you.

One idea, I use pg_search in my application to search for documents and I'm unable to use the searching algorithm with the newly encrypted db columns.

Any thoughts on getting this to work with the [pg_search](https://github.com/Casecommons/pg_search) gem or maybe gems like that one?

For example, in my model, I have: 

```
  encrypts :title, :body
  blind_index :title, :body

  pg_search_scope :kinda_spelled_like,
                  :against => [:title, :body],
                  :using => {
                    :tsearch => {
                      :prefix => true,
                      dictionary: "english"
                    }
                  }
```
But when I search, it says it can't find the "title" or "body" column. If I do a super basic query without using pg_search, it seems to query the table just fine - like you mentioned in your readme. But, I would like to use this pg_search gem because of the complex queries you can do with searching.

**Comment by ankane** [comment]
Hey @sgriff44, I'm not aware of a way to do efficient searching without leaking significant information about the data. If the number of records is small, you can load all records, decrypt them, and search in memory. If it's larger, you can use an expression blind index to narrow it down, and then search in memory (this approach leaks which records have the same value of the expression).

**Comment by ankane** [comment]
Just fyi, I've expanded on full text searching here: https://github.com/ankane/blind_index#like-ilike-and-fuzzy-searching

**Comment by eafif** [comment]
Great gem and very easy to get set up. How about dynamic `type`s for when the type is configurable based on the model instance?

```
encrypts :notes, type: -> { get_type }

def get_type
  # this will return one of the allowed types, such as :string, :integer, :float, etc...
end 
```

**Comment by ankane** [comment]
Hey @eafif, sorry for the delay. I don't think it's common enough to add.

Also, moving ideas to #155.
